src/Microsoft.DotNet.Watcher.Tools/DotNetWatcher.cs

```diff
@@ -1,176 +1,80 @@
-            }
-            if (dotnetArguments == null)
-            {
-                throw new ArgumentNullException(nameof(dotnetArguments));
-            }
-            if (cancellationToken == null)
-            {
-                throw new ArgumentNullException(nameof(cancellationToken));
-            }
+            Ensure.NotNull(processSpec, nameof(processSpec));
+            Ensure.NotNull(cancellationToken, nameof(cancellationToken));
 
-            var dotnetArgumentsAsString = ArgumentEscaper.EscapeAndConcatenateArgArrayForProcessStart(dotnetArguments);
-
-            var workingDir = Path.GetDirectoryName(projectFile);
+            var cancelledTaskSource = new TaskCompletionSource<object>();
+            cancellationToken.Register(state => ((TaskCompletionSource<object>)state).TrySetResult(null), cancelledTaskSource);
 
             while (true)
             {
-                await WaitForValidProjectJsonAsync(projectFile, cancellationToken);
-                cancellationToken.ThrowIfCancellationRequested();
+                var fileSet = await fileSetFactory.CreateAsync(cancellationToken);
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
 
                 using (var currentRunCancellationSource = new CancellationTokenSource())
                 using (var combinedCancellationSource = CancellationTokenSource.CreateLinkedTokenSource(
                     cancellationToken,
                     currentRunCancellationSource.Token))
+                using (var fileSetWatcher = new FileSetWatcher(fileSet))
                 {
-                    var fileWatchingTask = WaitForProjectFileToChangeAsync(projectFile, combinedCancellationSource.Token);
-                    var dotnetTask = WaitForDotnetToExitAsync(dotnetArgumentsAsString, workingDir, combinedCancellationSource.Token);
+                    var fileSetTask = fileSetWatcher.GetChangedFileAsync(combinedCancellationSource.Token);
+                    var processTask = _processRunner.RunAsync(processSpec, combinedCancellationSource.Token);
 
-                    var tasksToWait = new Task[] { dotnetTask, fileWatchingTask };
+                    var finishedTask = await Task.WhenAny(processTask, fileSetTask, cancelledTaskSource.Task).ConfigureAwait(false);
 
-                    int finishedTaskIndex = Task.WaitAny(tasksToWait, cancellationToken);
-
-                    // Regardless of the outcome, make sure everything is cancelled
+                    // Regardless of the which task finished first, make sure everything is cancelled
                     // and wait for dotnet to exit. We don't want orphan processes
                     currentRunCancellationSource.Cancel();
-                    Task.WaitAll(tasksToWait);
 
-                    cancellationToken.ThrowIfCancellationRequested();
+                    await Task.WhenAll(processTask, fileSetTask).ConfigureAwait(false);
```

Comment:
Yeah, that's the design. If process dies or is killed, we block on a file change to relaunch.